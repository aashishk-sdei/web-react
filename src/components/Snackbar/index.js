import React from 'react';
import Snackbar from '@material-ui/core/Snackbar';
import MuiAlert from '@material-ui/lab/Alert';
import { useDispatch, useSelector } from "react-redux";
import PropTypes from "prop-types";
import SnackbarContent from '@material-ui/core/SnackbarContent';
import Slide from '@material-ui/core/Slide';
import { removeMessage } from "../../redux/actions/toastr"
import { makeStyles } from '@material-ui/core/styles';
import { useTranslation } from "react-i18next";
import { tr } from "../../components/utils";

export const Alert = (props) => {
    return <MuiAlert elevation={6} variant="filled" {...props} />;
}
export const  SlideTransition = (props) => {
    return <Slide {...props} direction="up" />;
}
function getActionText(type){
  let text = "Undo"
  if(type === 'error') {
    text = "Try Again"
  }
  return text
}
function getActionClass(type) {
  let _class = 'text-green-1'
  if(type === 'error') {
    _class = 'text-maroon-1'
  } else if(type === 'info') {
    _class = 'text-skyblue-1'
  }
  return _class
}
const useStyles = makeStyles(() => ({
    mainRoot: {
      "&:nth-child(2)":{
        bottom:"80px"
      },
      "&:nth-child(3)":{
        bottom:"136px"
      }
    },
    root: {
      width: '100%',
      background: 'none',
      boxShadow: 'none',
      padding: 0,
      margin: 0
    },
    filledInfo: {
      background:"#212122"
    },
    filledSuccess: {
      background:"#388367"
    },
    filledError: {
      background:"#B02A4C"
    }
  }));
const SnackbarComponent = ({timing}) => {
  const { t } = useTranslation();
  const classes = useStyles();
  const transition = SlideTransition
  const dispatch = useDispatch()
  const { message: messageList } = useSelector((state) => {
    return state.toastr;
  });
  const handleClose = (message, reason) => {
    if (reason === 'clickaway') {
      return;
    }
    dispatch(removeMessage(message.key));
  };

  const handleExited = (message) => {
    dispatch(removeMessage(message.key));
  };
  const getAction = (message) => {
    const classesAction =  getActionClass(message.type)
    return <span  className={`${classesAction} ml-6 cursor-pointer hover:underline`} onClick={()=>{
      message.cta.action();
      handleExited(message);
    }}>{message.cta.text?message.cta.text:getActionText(message.type)}</span>
  }
  const getType = (message) => {
    let type = message.type
    if(message.type === 'warning') {
      type = 'error'
    } else if(message.type === 'error') {
      type = 'warning'
    }
    return type
  }
  return (
    <div className='multi-snkbr-toaster'>
    
    {messageList.map((message)=><Snackbar 
        key = {message.key}
        classes = {{
          root: classes.mainRoot
        }}
        anchorOrigin={{
            vertical: 'bottom',
            horizontal: 'left',
        }}
        autoHideDuration={timing}
        open={message!==null}
        onClose={(_event, reason)=>handleClose(message, reason)}
        onExited={()=>handleExited(message)}
        TransitionComponent={transition}>
        <SnackbarContent
            classes = {{
                root: classes.root
            }}
            message={message && <Alert 
                classes = {{
                  filledInfo:classes.filledInfo,
                  filledSuccess:classes.filledSuccess,
                  filledWarning:classes.filledError
                }}
                sx={{ width: '100%' }}
                {...![undefined, true].includes(message.icon)?{icon:message.icon}:{}}
                {...message.canClose?{onClose: handleClose.bind(null, message)}: {}}
                severity={getType(message)}>
                {message.url ? 
                  <span>
                    {tr(t, message.content)}
                    <a href={message.url} className='ml-4'>View</a>
                  </span> 
                  : <span>{tr(t, message.content)}</span>}
                {message.cta && getAction(message)}
            </Alert>}
        />
    </Snackbar>)}
    </div>);
}
SnackbarComponent.propTypes = {
    timing: PropTypes.number
}
SnackbarComponent.defaultProps = {
    timing: 6000
}
export default SnackbarComponent