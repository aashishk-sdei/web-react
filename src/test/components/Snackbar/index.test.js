import React from "react";
import Snackbar from "../../../components/Snackbar";
import { mount } from "enzyme";
import {useDispatch, useSelector} from "react-redux";
import { createShallow } from '@material-ui/core/test-utils';
const timing = 1000
jest.mock("react-redux", () => ({
    ...jest.requireActual("react-redux"),
    useSelector: jest.fn(),
    useDispatch: ()=>jest.fn()
}));
let shallowNew = createShallow(); 
describe("Snackbar Component", () => {
    const mockCallBack = jest.fn(()=> true);
    let snackBarComponent;
    beforeEach(() => {
        useSelector.mockImplementation(callback => {
            return callback({
                toastr: {
                    message: [
                        {
                            key: new Date().getTime()+'1', 
                            icon: true,
                            content: "Test Message", 
                            type: 'success', 
                            canClose: false,
                            url: false,
                            cta: {text: "Undo", action: mockCallBack}
                        },
                        {
                            key: new Date().getTime()+'2', 
                            icon: false,
                            content: "Test Message 2", 
                            type: 'error', 
                            canClose: true,
                            url: false
                        },
                        {
                            key: new Date().getTime()+'3', 
                            icon: false,
                            content: "Test Message 2", 
                            type: 'error', 
                            canClose: true,
                            url: false,
                            cta: {action: mockCallBack}
                        },
                        {
                            key: new Date().getTime()+'4', 
                            icon: false,
                            content: "Test Message 2", 
                            type: 'warning', 
                            canClose: true,
                            url: "http://google.com",
                            cta: {action: mockCallBack}
                        },
                        {
                            key: new Date().getTime()+'5', 
                            icon: false,
                            content: "Test Message 2", 
                            type: 'info', 
                            canClose: true,
                            url: "http://google.com",
                            cta: {action: mockCallBack}
                        }
                    ]
                }
            });
          });
        snackBarComponent = mount(<Snackbar timing={timing}/>)
    });
    it('match add Message', () => {
        expect(snackBarComponent).toMatchSnapshot();
    });
    it('Snackbar event onClick span', ()=>{
        snackBarComponent.find('span.ml-6.cursor-pointer').at(0).simulate('click');
    })
    afterEach(() => {
        useSelector.mockClear();
    });
});