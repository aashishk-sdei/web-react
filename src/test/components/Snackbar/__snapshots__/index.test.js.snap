// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snackbar Component match add Message 1`] = `ReactWrapper {}`;
